STM32/I2C/Src/sm4291.c

```diff
@@ -0,0 +1,310 @@
+
+        void* ret = malloc(sizeof(temp));
+
+        if(ret) {
+            ret = memcpy(ret, &temp, sizeof(temp));
+        }
+
+        return (sm4291_i2c_handle_t*) ret;
+    }
+    else {
+        return NULL;
+    }
+}
+
+/*!
+** @brief Get the temperature from the sensor
+*/
+int sm4291GetTemp(sm4291_i2c_handle_t* i2c, double* result) {
+    int16_t temp;
+    if(0 != sm4291ReadReg(i2c, ADDR_DSP_T, (uint16_t*)&temp)) {
+        return -1;
+    };
+
+    *result = temp * TEMP_ADC_SCALAR + TEMP_ADC_OFFSET;
+    return 0;
+}
+
+/*!
+** @brief Get the pressure from the sensor
+*/
+int sm4291GetPressure(sm4291_i2c_handle_t* i2c, double* result) {
+    assert_param(i2c);
+
+    int16_t press;
+    if(0 != sm4291ReadReg(i2c, ADDR_DSP_P, (uint16_t*)&press)) {
+        return -1;
+    };
+
+    *result = press * i2c->press_scaler + i2c->press_offset;
+    return 0;
+}
+
+/*!
+** @brief Get the serial number from the sensor
+*/
+int sm4291GetSerial(sm4291_i2c_handle_t* i2c, uint32_t* result) {
+    uint16_t ser0;
+    uint16_t ser1;
+
+    if(0 != sm4291ReadReg(i2c, ADDR_SER0, &ser0)) {
+        return -1;
+    };
+
+    if(0 != sm4291ReadReg(i2c, ADDR_SER1, &ser1)) {
+        return -2;
+    };
+
+    *result = (((uint32_t)ser1 << 16U) & 0xFFFF0000U) | ((uint32_t)ser0 & 0xFFFFU);
+    return 0;
+}
+
+/*!
+** @brief Get the status register
+*/
+int sm4291GetStatus(sm4291_i2c_handle_t* i2c, uint16_t* status) {
+    return sm4291ReadReg(i2c, ADDR_STATUS, status);
+}
+
+/*!
+** @brief Get the status sync register
+*/
+int sm4291GetStatusSync(sm4291_i2c_handle_t* i2c, uint16_t* status) {
+    return sm4291ReadReg(i2c, ADDR_STATUS_SYNC, status);
+}
+
+/*!
+** @brief Reset the device
+*/
+int sm4291Reset(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_RESET);
+}
+
+/*!
+** @brief Put the device into sleep mode
+**
+** @note The datasheet doesn't specify how t get the device out of sleep mode
+*/
+int sm4291Sleep(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_SLEEP);
+}
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Reads a 16-bit register from the device
+*/
+static int sm4291ReadReg(sm4291_i2c_handle_t* i2c, uint8_t addr, uint16_t* result) {
+    assert_param(i2c);
+
+    uint16_t dev_addr = i2c->crc ? TEMP_I2C_CRC_ADDR : TEMP_I2C_NO_CRC_ADDR;
+
+    if(i2c->crc) {
+        uint8_t crc4_data[2U] = {(uint8_t)(addr >> 4U), ((addr & 0xFU) << 4U) & 0x1U};
+        uint8_t crc4 = genCrc4(crc4_data, 2U);
+
+        uint8_t buf[2U] = {addr, (uint8_t)(0x10U | crc4)};
+        if(HAL_OK != HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, buf, 2, 1)) {
+            return -1;
+        };
+
+        uint8_t buf1[3U] = {0};
+        if(HAL_OK != HAL_I2C_Master_Receive(i2c->i2c, dev_addr << 1, buf1, 3, 1)) {
```

Comment:
Let's change the buffer size inputs in `HAL_I2C...(...)`  to have the U suffix to match the rest of the file.